### PR TITLE
Document Lifespan state

### PR DESCRIPTION
**Pull request by Kludex**
- This is a continuation of https://github.com/encode/starlette/pull/2060.

### CODE REVIEW

File: docs/events.md [hunk 1]
```diff
@@ -62,6 +62,62 @@ app = Starlette(routes=routes, lifespan=lifespan)
...
+## Lifespan State
+
+The event handlers also accept a `state` argument, which is a dictionary
+that can be used to share the objects between the startup and shutdown handlers,
+and the requests.
+
+```python
+import httpx
+from starlette.applications import Starlette
+from starlette.routing import Route
```
Comment (Kludex) [inline docs/events.md:75]:
```suggestion
from starlette.responses import PlainTextResponse
from starlette.routing import Route
```

File: docs/events.md [hunk 2]
```diff
@@ -62,6 +62,62 @@ app = Starlette(routes=routes, lifespan=lifespan)
...
+```python
+import contextlib
+import httpx
+from starlette.applications import Starlette
+from starlette.routing import Route
+
+
+@contextlib.asynccontextmanager
+async def lifespan(app, state):
+    state["http_client"] = httpx.AsyncClient()
```
Comment (abersheeran) [inline docs/events.md]: Why not use `async with`? 🤔️

Comment (adriangb) [reply docs/events.md]: +1

Comment (Kludex) [reply docs/events.md]: oh... :eyes:

Comment (Kludex) [reply docs/events.md]: done

**Review by abersheeran — APPROVED** [review]

File: docs/events.md [hunk 3]
```diff
@@ -62,6 +62,63 @@ app = Starlette(routes=routes, lifespan=lifespan)
...
+        yield
+
+
+app = Starlette(
+    lifespan=lifespan,
+    routes=[Route("/", homepage)]
+)
+```
+
+The `state` received on the requests is a copy of the state received on the
```
Comment (adriangb) [inline docs/events.md]:
```suggestion
The `state` received on the requests is a **shallow** copy of the state received on the
```

**Review by adriangb — APPROVED** [review]

**Pull request merged**